Add protocol and transport documentation (#663)
python-kasa/python-kasa: docs/source/design.rst

    If you only need some module-specific information,
    you can call the wanted method on the module to avoid using :meth:`~kasa.SmartDevice.update`.

Protocols and Transports
************************

The library supports two different TP-Link protocols, ``IOT`` and ``SMART``.

> ``IOT`` is the original Kasa protocol and ``SMART`` is the newer protocol supported by TAPO devices and newer KASA devices.

Reviewer: Probably worth mentioning here, that the protocols imply different set of commands? In the future, we can improve the module docs on that, too.

Author: Done

Confusingly TP-Link originally called the Kasa line "Kasa Smart" and hence this library used "Smart" in a lot of the
module and class names but actually they were built to work with the ``IOT`` protocol.

In 2021 TP-Link started updating the underlying communication transport used by Kasa devices to make them more secure.
It switched from a TCP connection with static XOR type of encryption to a transport called ``KLAP`` which communicates

> over http and uses handshakes to negotiate a dynamic encrpytion cipher.

Reviewer: ```suggestion
over http and uses handshakes to negotiate a dynamic encryption cipher.
```

Author: Done

This automatic update was put on hold and only seemed to affect UK HS100 models.

In 2023 TP-Link started updating the underlying communication transport used by Tapo devices to make them more secure.
It switched from AES encryption via public key exchange to use ``KLAP`` encryption and negotiation due to concerns
around impersonation with AES.
The encryption cipher is the same as for Kasa KLAP but the handshake seeds are slightly different.
Also in 2023 TP-Link started releasing newer Kasa branded devices using the ``SMART`` protocol.
This appears to be driven by hardware version rather than firmware.

> In order to support these different configurations the library is migrating from a single :class:`~TPLinkSmartHomeProtocol`
> that support XOR encryption and the IOT protocol to seperate transports and protocols.

Reviewer: ```suggestion
In order to support these different configurations the library migrated from a single :class:`~TPLinkSmartHomeProtocol` to support pluggable transports and protocols.
```
Maybe something like this?

Author: Done
